productivity-suite/piercing-library/src/message-bus/message-bus.ts

```diff
@@ -0,0 +1,51 @@
+export type MessageBusCallback = (value: any) => void;
+
+export type MessageBusState = Record<string, any>;
+
+export interface MessageBus {
+  state: MessageBusState;
+  dispatch(eventName: string, value: any): void;
+  listen(
+    eventName: string,
+    callback: (value: any) => void
+  ): (() => void) | null;
+  latestValue(eventName: string): any;
+}
+
+export class GenericMessageBus implements MessageBus {
+  protected _callbacksMap: Map<string, MessageBusCallback[]> = new Map();
+
+  constructor(protected _state: MessageBusState = {}) {}
+
+  get state(): MessageBusState {
+    return JSON.parse(JSON.stringify(this._state));
+  }
+
+  dispatch(eventName: string, value: any) {
+    this._state[eventName] = value;
+    const callbacksForEvent = this._callbacksMap.get(eventName) ?? [];
+    callbacksForEvent.forEach((callback) => callback(value));
+  }
+
+  listen(eventName: string, callback: MessageBusCallback) {
+    const latestValue = this.latestValue(eventName);
+    if (latestValue) {
```

Comment:
With the addition of the `MessageBusValue` I changed this check to be `latestValue !== undefined` since undefined is not a valid `MessageBusValue`:
![Screenshot 2022-11-02 at 11 44 04](https://user-images.githubusercontent.com/61631103/199481717-13ab528e-bd89-4516-b403-1c5859e6b6d6.png)

I hope that's ok? 🙂